### PR TITLE
Added different suppression calculation for explosives and explosions

### PR DESCRIPTION
**Pull request by AluminumAlman**
## Changes

- Explosions and explosive projectiles use a different formula for giving suppression; they also ignore sharp armor;
- Explosions deal more suppression than the explosive projectile itself.

## Reasoning

- Explosion suppression was broken due to the fact that it always checked against average sharp armor with its sharp penetration, which is 0. This should fix the issue of explosives not doing enough suppression, while raising the T in the lerp to the power of two should help reduce suppression the further a pawn is from the projectile or explosion;
- Projectiles can suppress the same pawn multiple times if it is in the suppression radius of the projectile for multiple ticks, therefore to me it didn't make sense how explosive projectiles can suppress more than the explosion itself.

## Testing

- [x] Compiles without warnings;
- [x] Game runs without errors;
- [ ] (For compatibility patches) ...with and without patched mod loaded;
- [ ] Playtested a colony;
- Did battlefield testing (image taken with the fragment changes included) (10000 point raid, one doomsday rocket);
image

### CODE REVIEW

**Comment by AluminumAlman** [comment]
Additionally, I was wondering whether or not to reduce suppression from explosives while they're flying trough the air, because it would appear that explosive projectiles can be more suppressive before the explosion even occurs.

File: Source/CombatExtended/CombatExtended/Projectiles/ProjectileCE.cs [hunk 1]
```diff
@@ -870,9 +870,18 @@ private void ApplySuppression(Pawn pawn)
             {
                 suppressionAmount = def.projectile.GetDamageAmount(1);
                 var propsCE = def.projectile as ProjectilePropertiesCE;
-                var penetrationAmount = propsCE?.armorPenetrationSharp ?? 0f;
-                var armorMod = penetrationAmount <= 0 ? 0 : 1 - Mathf.Clamp(pawn.GetStatValue(CE_StatDefOf.AverageSharpArmor) * 0.5f / penetrationAmount, 0, 1);
-                suppressionAmount *= armorMod;
+                var explodeRadius = propsCE.explosionRadius;
+                if (explodeRadius < 0.1f)
```
Comment (AluminumAlman) [inline Source/CombatExtended/CombatExtended/Projectiles/ProjectileCE.cs:874]: This could be a possible nit-pick for someone, too.

File: Source/CombatExtended/CombatExtended/Projectiles/ProjectileCE.cs [hunk 2]
```diff
@@ -1105,7 +1115,7 @@ public virtual void Impact(Thing hitThing)
                 }
 
                 foreach (var thing in suppressThings)
-                    ApplySuppression(thing);
+                    ApplySuppression(thing, 3f);
```
Comment (AluminumAlman) [inline Source/CombatExtended/CombatExtended/Projectiles/ProjectileCE.cs:1118]: I can either keep this value at `3f` or reduce it, as I am unsure of how balanced it is.

Comment (perkinslr) [reply Source/CombatExtended/CombatExtended/Projectiles/ProjectileCE.cs:1117]: It's usually best to make "magic constants" use constant variables at the top of the class (for examples, look at the top of ProjectileCE).  

So it would read `ApplySuppression(thing, ExplosionSuppression);`

File: Source/CombatExtended/CombatExtended/StatWorkers/StatWorker_ArmorCoverage.cs [hunk 3]
```diff
@@ -24,7 +24,7 @@ public override float GetValueUnfinalized(StatRequest req, bool applyPostProcess
                 totalCoverage += coverage;
             }
 
-            return totalCoverage > 0 ? weightedArmor / totalCoverage : 0;
+            return totalCoverage > 0 ? weightedArmor : 0;
```
Comment (perkinslr) [inline Source/CombatExtended/CombatExtended/StatWorkers/StatWorker_ArmorCoverage.cs:27]: What's the reason for this change?  This is likely to have ripple effects elsewhere.

File: Source/CombatExtended/CombatExtended/Projectiles/ProjectileCE.cs [hunk 4]
```diff
@@ -870,9 +870,20 @@ private void ApplySuppression(Pawn pawn)
...
+                    var penetrationAmount = propsCE?.armorPenetrationSharp ?? 0f;
+                    var armorMod = penetrationAmount <= 0 ? 0 : 1 - Mathf.Clamp(pawn.GetStatValue(CE_StatDefOf.AverageSharpArmor) * 0.5f / penetrationAmount, 0, 1);
+                    suppressionAmount *= armorMod;
+                }
+                else
+                {
+                    var dPosX = ExactPosition.x - pawn.DrawPos.x;
+                    var dPosZ = ExactPosition.z - pawn.DrawPos.z;
+                    // Affected by the ratio of distance from the explosion to the max suppression radius raised to the power of two. Larger suppression amount compared to linear interpolation. Also helps that square root isn't used
+                    suppressionAmount *= Mathf.Clamp01(1f - (dPosX * dPosX + dPosZ * dPosZ) / ((explodeRadius + SuppressionRadius) * (explodeRadius + SuppressionRadius))) * explosiveSuppressionMultiplier;
```
Comment (perkinslr) [inline Source/CombatExtended/CombatExtended/Projectiles/ProjectileCE.cs:885]: If you are concerned this is a hot function and want to optimize it, you should store `explodeRadius + SuppressionRadius` in a temporary local to avoid the repeated addition.  In practice, since SuppressionRadius is `const`, the JIT *should* take care of it, but worth keeping in mind.

**Review by perkinslr — CHANGES_REQUESTED** [review]

File: Source/CombatExtended/CombatExtended/StatWorkers/StatWorker_ArmorCoverage.cs (hunk 3 above)
Comment (AluminumAlman) [reply Source/CombatExtended/CombatExtended/StatWorkers/StatWorker_ArmorCoverage.cs:27]: This caused pawns to believe that they have more armor coverage than they actually did under some circumstances, for example, wearing ONLY an armor vest would result in having 8mm armor coverage.

Comment (perkinslr) [reply Source/CombatExtended/CombatExtended/StatWorkers/StatWorker_ArmorCoverage.cs:27]: Ah, that makes sense.  If we assume total coverage is 1, since the sum of all body parts coverage *should* be one, then we can drop the division.

That said, the reason for the `totalCoverage > 0` was to avoid a divide-by-zero.  Without the division, `totalCoverage` can go completely away.

File: Source/CombatExtended/CombatExtended/Projectiles/ProjectileCE.cs [hunk 5]
```diff
@@ -1092,20 +1109,18 @@ public virtual void Impact(Thing hitThing)
...
 
                     if (explodePos.y < SuppressionRadius)
-                        suppressThings.AddRange(GenRadial
-                            .RadialDistinctThingsAround(explodePos.ToIntVec3(), Map, SuppressionRadius + (explodingComp.props as CompProperties_ExplosiveCE).explosiveRadius, true).OfType<Pawn>());
+                        suppressThings.AddRange(explodePos.ToIntVec3().PawnsInRange(Map, SuppressionRadius + (explodingComp.props as CompProperties_ExplosiveCE).explosiveRadius));
                 }
 
                 foreach (var thing in suppressThings)
-                    ApplySuppression(thing);
+                    ApplySuppression(thing, explosionSuppressionFactor);
```
Comment (perkinslr) [inline Source/CombatExtended/CombatExtended/Projectiles/ProjectileCE.cs:1123]: Just like `SuppressionRadius`, this should be `TitleCase`.  Not that CE is super consistent with that, so not a *huge* deal.

**Review by perkinslr — APPROVED** [review]

File: Source/CombatExtended/CombatExtended/Projectiles/ProjectileCE.cs [hunk 6]
```diff
@@ -27,6 +27,11 @@ public abstract class ProjectileCE : ThingWithComps
         /// </summary>
         private const int collisionCheckSize = 5;
 
+        /// <summary>
+        /// Additional suppression multiplier for suppression caused by an explosion.
+        /// </summary>
+        private const float explosionSuppressionFactor = 3f;
```
Comment (AluminumAlman) [inline Source/CombatExtended/CombatExtended/Projectiles/ProjectileCE.cs:33]: I'm unsure of the balance implications this factor has.

Comment (ItsSamGray) [reply Source/CombatExtended/CombatExtended/Projectiles/ProjectileCE.cs:33]: Lower it to 2f if you think it's too high, gotta be aware of not nerfing it to near the previous state.

**Comment by Breadbox2k19** [comment]
> Additionally, I was wondering whether or not to reduce suppression from explosives while they're flying trough the air, because it would appear that explosive projectiles can be more suppressive before the explosion even occurs.

Yes please, infact I should say explosive projectile should do little to no surpression until they explode, stickbombs should not pin people down until they explode.

**Review by N7Huntsman — APPROVED** [review]
Works in testing. Good to go.
